src/Utilities/Compiler/HashUtilities.cs

```diff
@@ -1,63 +1,76 @@
+#endif
 
 namespace Analyzer.Utilities
 {
     internal static class HashUtilities
     {
         internal static int GetHashCodeOrDefault(this object? objectOpt) => objectOpt?.GetHashCode() ?? 0;
 
-        internal static int Combine(int newKey, int currentKey)
+        internal static int Combine<T>(ImmutableArray<T> array)
         {
-            return unchecked((currentKey * (int)0xA5555529) + newKey);
+            var hashCode = new RoslynHashCode();
+            Combine(array, ref hashCode);
+            return hashCode.ToHashCode();
         }
 
-        internal static int Combine<T>(ImmutableArray<T> array) => Combine(array, 0);
-        internal static int Combine<T>(ImmutableArray<T> array, int currentKey) => Combine(array, array.Length, currentKey);
-
-        public static int Combine<T>(params T[] sequence)
-            => Combine(sequence, sequence.Length, currentKey: 0);
-
-        public static int Combine<T>(IEnumerable<T> sequence, int length, int currentKey)
+        internal static void Combine<T>(ImmutableArray<T> array, ref RoslynHashCode hashCode)
         {
-            var hashCode = Combine(length, currentKey);
-            foreach (var element in sequence)
+            foreach (var element in array)
             {
-                hashCode = Combine(element.GetHashCodeOrDefault(), hashCode);
+                hashCode.Add(element);
             }
-
-            return hashCode;
         }
 
-        internal static int Combine<T>(ImmutableStack<T> stack) => Combine(stack, 0);
-        internal static int Combine<T>(ImmutableStack<T> stack, int currentKey)
+        internal static int Combine<T>(ImmutableStack<T> stack)
         {
-            var hashCode = currentKey;
+            var hashCode = new RoslynHashCode();
+            Combine(stack, ref hashCode);
+            return hashCode.ToHashCode();
+        }
 
-            var stackSize = 0;
+        internal static void Combine<T>(ImmutableStack<T> stack, ref RoslynHashCode hashCode)
+        {
             foreach (var element in stack)
             {
-                hashCode = Combine(element.GetHashCodeOrDefault(), hashCode);
-                stackSize++;
+                hashCode.Add(element);
             }
+        }
 
-            return Combine(stackSize, hashCode);
+        internal static int Combine<T>(ImmutableHashSet<T> set)
+        {
+            var hashCode = new RoslynHashCode();
+            Combine(set, ref hashCode);
+            return hashCode.ToHashCode();
         }
 
-        internal static int Combine<T>(ImmutableHashSet<T> set) => Combine(set, 0);
-        internal static int Combine<T>(ImmutableHashSet<T> set, int currentKey)
-            => Combine(set.Select(element => element.GetHashCodeOrDefault()).Order(),
-                       set.Count,
-                       currentKey);
+        internal static void Combine<T>(ImmutableHashSet<T> set, ref RoslynHashCode hashCode)
+        {
+            foreach (var element in set)
```

Comment:
Is enumerating a hash set (and dictionary) with the same elements always going to have a stable order? We used to sort the elements' hash codes before combining.

Reply:
`ImmutableHashSet<T>` internally sorts elements by hash code. This is not the same as other collections but works for what we need here.